test/multi_rollup_sequencer.go

```diff
@@ -0,0 +1,134 @@
+package test
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/rollkit/go-sequencing"
+)
+
+// MultiRollupSequencer is a sequencer for testing that serves multiple rollups
+type MultiRollupSequencer struct {
+	rollups      map[string]*RollupData
+	rollupsMutex sync.RWMutex
+}
+
+// RollupData holds the data for a specific rollup, including its transaction queue, last batch hash, and seen batches.
+type RollupData struct {
+	tq                 *TransactionQueue
+	lastBatchHash      []byte
+	lastBatchHashMutex sync.RWMutex
+
+	seenBatches      map[string]struct{}
+	seenBatchesMutex sync.Mutex
+}
+
+// SubmitRollupTransaction implements sequencing.Sequencer.
+func (d *MultiRollupSequencer) SubmitRollupTransaction(ctx context.Context, req sequencing.SubmitRollupTransactionRequest) (*sequencing.SubmitRollupTransactionResponse, error) {
+	rollup, err := d.getOrCreateRollup(req.RollupId)
+	if err != nil {
+		return nil, err
+	}
+	rollup.tq.AddTransaction(req.Tx)
+	return &sequencing.SubmitRollupTransactionResponse{}, nil
+}
+
+// GetNextBatch implements sequencing.Sequencer.
+func (d *MultiRollupSequencer) GetNextBatch(ctx context.Context, req sequencing.GetNextBatchRequest) (*sequencing.GetNextBatchResponse, error) {
+	rollup, err := d.getOrCreateRollup(req.RollupId)
+	if err != nil {
+		return nil, err
+	}
+
+	now := time.Now()
+	rollup.lastBatchHashMutex.RLock()
+	lastBatchHash := rollup.lastBatchHash
+	rollup.lastBatchHashMutex.RUnlock()
+
+	if lastBatchHash == nil && req.LastBatchHash != nil {
+		return nil, errors.New("lastBatch is supposed to be nil")
+	} else if lastBatchHash != nil && req.LastBatchHash == nil {
+		return nil, errors.New("lastBatch is not supposed to be nil")
+	} else if !bytes.Equal(lastBatchHash, req.LastBatchHash) {
+		return nil, errors.New("supplied lastBatch does not match with sequencer last batch")
+	}
```

Comment:
Could you simplify this to just use [DeepEqual](https://pkg.go.dev/reflect#DeepEqual) and then in the error message display the values to illustrate the difference?

Reply:
Same would apply to the Dummy Sequencer https://github.com/rollkit/go-sequencing/pull/18/files#diff-fbe51e832ecc1c24b1b747b8b363fed13b0b160d55035638c0ea9d9c9d4d13bdR108

Reply:
not sure how deep equal solves my issue. the checks have to be specific here in that, if either of the params are nil and other is not then error. if both are non-nil then check for bytes. i think this should be enough?
```
if (lastBatchHash == nil && req.LastBatchHash != nil) || (lastBatchHash != nil && req.LastBatchHash == nil) {
    return nil, fmt.Errorf("nil mismatch: lastBatchHash = %v, req.LastBatchHash = %v", lastBatchHash, req.LastBatchHash)
} else if lastBatchHash != nil && !bytes.Equal(lastBatchHash, req.LastBatchHash) {
    return nil, fmt.Errorf("batch hash mismatch: lastBatchHash = %x, req.LastBatchHash = %x", lastBatchHash, req.LastBatchHash)
}
```

Reply:
you are checking for equality essentially right? both nil, both not nil, or bytes are equal. That is what deep equal is checking. 

> Slice values are deeply equal when all of the following are true: they are both nil or both non-nil, they have the same length, and either they point to the same initial entry of the same underlying array (that is, &x[0] == &y[0]) or their corresponding elements (up to length) are deeply equal. Note that a non-nil empty slice and a nil slice (for example, []byte{} and []byte(nil)) are not deeply equal.